### PR TITLE
Parser support for assigns contracts on loops

### PR DESCRIPTION
**Pull request by aalok-thakkar**
1. Created cprover_loop_contract that has cprover_contract_assigns_opt and loop_invariant_opt.

This adds support for parsing assigns clauses for loops.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/parser.y [hunk 1]
```diff
@@ -3271,6 +3271,12 @@ cprover_function_contract_sequence_opt:
         | cprover_function_contract_sequence
         ;
 
+cprover_loop_contract:
+          cprover_contract_assigns_opt
+          loop_invariant_opt
+        | loop_invariant_opt
```
Comment (SaswatPadhi) [inline src/ansi-c/parser.y]: Why do we need these two rules? Wouldn't the first rule suffice? `cprover_contract_assigns_opt` can expand to epsilon.

File: src/ansi-c/parser.y [hunk 2]
```diff
@@ -3271,6 +3271,12 @@ cprover_function_contract_sequence_opt:
         | cprover_function_contract_sequence
         ;
 
+cprover_loop_contract:
+          cprover_contract_assigns_opt
+          loop_invariant_opt
```
Comment (SaswatPadhi) [inline src/ansi-c/parser.y]: Could you please rename `loop_invariant_opt` to `cprover_contract_invariant_opt` for consistency? 

And, similarly, `cprover_contract_decreases_opt` when you resolve merge conflicts.

In the future, I would update the invariant keyword to `__CPROVER_invariant`, and we can specify both decreases and invariants in contracts for (recursive) functions as well.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Please rebase against current `develop` and resolve the merge conflict -- it's due to the decreases contract that Long's PR introduced. That should be added to `cprover_loop_contract` and should be removed from individual iteration statements.

File: src/ansi-c/parser.y (hunk 1 above)
Comment (aalok-thakkar) [reply src/ansi-c/parser.y]: I have now created cprover_contract_assigns and cprover_contract_assigns_opt. This resolves the issue.

File: src/ansi-c/parser.y (hunk 2 above)
Comment (aalok-thakkar) [reply src/ansi-c/parser.y]: Renamed, and merge conflicts resolved.

**Review by feliperodri — APPROVED** [review]
LGTM.

**Comment by SaswatPadhi** [comment]
Marking this as draft. There are several issues with commits, including merge conflicts.
@aalok-thakkar please fix these.

**Review by martin-cs — COMMENTED** [review]
Please rebase as there seems to be at least one unrelated commit in this PR.

File: src/ansi-c/parser.y [hunk 3]
```diff
@@ -3552,4 +3564,4 @@ parameter_postfix_abstract_declarator:
         }
         ;
 
-%%
+%%
```
Comment (SaswatPadhi) [inline src/ansi-c/parser.y]: Please add this newline back.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
LGTM, but one minor issue.

File: src/ansi-c/parser.y (hunk 3 above)
Comment (aalok-thakkar) [reply src/ansi-c/parser.y]: Added.

**Review by SaswatPadhi — APPROVED** [review]

**Review by martin-cs — CHANGES_REQUESTED** [review]
The patch looks OK but I really don't think we should be adding syntax extensions without test cases that at least demonstrate how to use them.

File: regression/contracts/invar_loop_constant_pass/test.desc [hunk 4]
```diff
@@ -10,4 +10,4 @@ main.c
 ^VERIFICATION SUCCESSFUL$
 --
 This test checks that the invariant is correctly used to satisfy
-all assertions that follow a loop.
+all assertions that follow a loop.
```
Comment (feliperodri) [inline regression/contracts/invar_loop_constant_pass/test.desc]: Fix end of line.

File: regression/contracts/invar_dynamic_struct_member/main.c [hunk 5]
```diff
@@ -9,7 +9,7 @@ void main()
   t->x = 0;
 
   unsigned n;
-  for(unsigned i = 0; i < n; ++i)
+  for(unsigned i = 0; i < n; ++i)
```
Comment (feliperodri) [inline regression/contracts/invar_dynamic_struct_member/main.c]: Remove white space.

File: regression/contracts/invar_check_multiple_loops/main.c [hunk 6]
```diff
@@ -26,4 +27,4 @@ int main()
   assert(x == y + 2 * n);
 
   return 0;
-}
+}
```
Comment (feliperodri) [inline regression/contracts/invar_check_multiple_loops/main.c]: Fix end of line.

File: regression/contracts/invar_assigns_opt/test.desc [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+^\[main.assertion.1\] .* assertion r1 == 0: SUCCESS$
+^\[main.assertion.2\] .* assertion s1 == 1: SUCCESS$
+^\[main.3\] .* Check loop invariant before entry: SUCCESS$
+^\[main.4\] .* Check that loop invariant is preserved: SUCCESS$
+^\[main.assertion.3\] .* assertion r2 == 0: SUCCESS$
+^\[main.assertion.4\] .* assertion s2 == 1: SUCCESS$
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that adding assigns clause is optional
```
Comment (feliperodri) [inline regression/contracts/invar_assigns_opt/test.desc:17]: Could you rebase this branch and add a test case using an empty assigns clause (e.g., `__CPROVER_assigns()`)?

File: regression/contracts/invar_assigns_opt/main.c [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    // clang-format on
+    {
+      s2 = 1;
+      r2--;
+    }
+  assert(r2 == 0);
+  assert(s2 == 1);
+
+  return 0;
+
```
Comment (feliperodri) [inline regression/contracts/invar_assigns_opt/main.c]: Remove unnecessary empty spaces.

File: regression/contracts/invar_assigns_alias_analysis/test.desc [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+^EXIT=0$
+^SIGNAL=0$
+^\[main.1\] .* Check loop invariant before entry: SUCCESS$
+^\[main.2\] .* Check that loop invariant is preserved: SUCCESS$
+^\[main.assertion.1\] .* assertion b->data[5] == 0: FAILURE$
+^VERIFICATION FAILED$
+--
+--
+This test shows a known bug in the current implementation
+for assigns clause. Fixing this is a part of future work.
```
Comment (feliperodri) [inline regression/contracts/invar_assigns_alias_analysis/test.desc]: Describe the bug.

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: regression/contracts/invar_assigns_alias_analysis/test.desc [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+^EXIT=0$
+^SIGNAL=0$
+^\[main.1\] .* Check loop invariant before entry: SUCCESS$
+^\[main.2\] .* Check that loop invariant is preserved: SUCCESS$
+^\[main.assertion.1\] .* assertion b->data[5] == 0: FAILURE$
+^VERIFICATION FAILED$
+--
+--
+This test shows that the current implementation only supports
+parsing and still uses alias analysis. Fixing this is a part
```
Comment (feliperodri) [inline regression/contracts/invar_assigns_alias_analysis/test.desc]: What do you mean by "current implementation only supports parsing and still uses alias analysis"? Could you elaborate?

File: regression/contracts/invar_assigns_empty/test.desc [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+--apply-loop-contracts
+^EXIT=0$
+^SIGNAL=0$
+^\[main.1\] .* Check loop invariant before entry: SUCCESS$
+^\[main.2\] .* Check that loop invariant is preserved: SUCCESS$
+^\[main.assertion.1\] .* assertion r == 0: SUCCESS$
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that empty assigns clause is supported.
```
Comment (feliperodri) [inline regression/contracts/invar_assigns_empty/test.desc]:
```
This test checks that empty assigns clause is supported in loop contracts.
```

File: regression/contracts/invar_check_multiple_loops/main.c (hunk 6 above)
Comment (feliperodri) [reply regression/contracts/invar_check_multiple_loops/main.c]: ?

File: regression/contracts/invar_dynamic_struct_member/main.c (hunk 5 above)
Comment (feliperodri) [reply regression/contracts/invar_dynamic_struct_member/main.c]: ?

File: regression/contracts/invar_loop_constant_pass/test.desc (hunk 4 above)
Comment (feliperodri) [reply regression/contracts/invar_loop_constant_pass/test.desc]: ?

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: regression/contracts/invar_assigns_alias_analysis/test.desc (hunk 10 above)
Comment (aalok-thakkar) [reply regression/contracts/invar_assigns_alias_analysis/test.desc]: This PR only adds parsing to it, and assertion statements are yet not generated. Therefore, this proof does not go through.

Comment (SaswatPadhi) [reply regression/contracts/invar_assigns_alias_analysis/test.desc]: What Aalok is saying is that, he added this test just to check the parsing support (and address Martin's comment regarding usage of assigns clause). But it's not considered during instrumentation, so this test which he obtained from my bug report, still fails as before.

**Review by feliperodri — APPROVED** [review]
LGTM.

**Comment by feliperodri** [comment]
@martin-cs could you give another review on this PR?

File: regression/contracts/invar_assigns_alias_analysis/test.desc [hunk 12]
```diff
@@ -0,0 +1,14 @@
+KNOWNBUG
```
Comment (martin-cs) [inline regression/contracts/invar_assigns_alias_analysis/test.desc:1]: If it is marked as a known bug then the test will not be run.  Is it possible to either have just a test for the parsing or test for the "incorrect" results until it is implemented.

**Review by martin-cs — APPROVED** [review]
Sure @feliperodri .  There are now tests so no longer blocking.

File: regression/contracts/invar_assigns_alias_analysis/test.desc (hunk 12 above)
Comment (feliperodri) [reply regression/contracts/invar_assigns_alias_analysis/test.desc:1]: @SaswatPadhi @aalok-thakkar

Comment (SaswatPadhi) [reply regression/contracts/invar_assigns_alias_analysis/test.desc:1]: There is one such test already, to check just the parsing. Please see: regression/contracts/invar_assigns_opt/main.c.

Comment (martin-cs) [reply regression/contracts/invar_assigns_alias_analysis/test.desc:1]: Thanks.

**Review by chrisr-diffblue — APPROVED** [review]
